cli_tools/gce_windows_upgrade/upgrader/validators.go

```diff
@@ -0,0 +1,180 @@
+
+	// Update 'project' value for logging purpose
+	*u.ProjectPtr = u.project
+
+	return nil
+}
+
+func validateOSVersion(sourceOS, targetOS string) error {
+	if sourceOS == "" {
+		return daisy.Errf("Flag -source-os must be provided. Please choose a supported version from {%v}.", strings.Join(SupportedSourceOSVersions(), ", "))
+	}
+	if _, ok := supportedSourceOSVersions[sourceOS]; !ok {
+		return daisy.Errf("Flag -source-os value '%v' unsupported. Please choose a supported version from {%v}.", sourceOS, strings.Join(SupportedSourceOSVersions(), ", "))
+	}
+	if targetOS == "" {
+		return daisy.Errf("Flag -target-os must be provided. Please choose a supported version from {%v}.", strings.Join(SupportedTargetOSVersions(), ", "))
+	}
+	if _, ok := supportedTargetOSVersions[targetOS]; !ok {
+		return daisy.Errf("Flag -target-os value '%v' unsupported. Please choose a supported version from {%v}.", targetOS, strings.Join(SupportedTargetOSVersions(), ", "))
+	}
+	// We may chain several upgrades together in the future (for example, 2008r2->2012r2->2016->2019).
+	// For now, we only support one-step upgrade.
+	if expectedTargetOS, _ := supportedSourceOSVersions[sourceOS]; expectedTargetOS != targetOS {
+		return daisy.Errf("Can't upgrade from %v to %v. Can only upgrade to %v.", sourceOS, targetOS, expectedTargetOS)
+	}
+	return nil
+}
+
+func validateInstanceURI(instanceURI string, derivedVars *derivedVars) error {
+	if instanceURI == "" {
+		return daisy.Errf("Flag -instance must be provided")
+	}
+	m := daisy.NamedSubexp(instanceURLRgx, instanceURI)
+	if m == nil {
+		return daisy.Errf("Please provide the instance flag in the form of 'projects/<project>/zones/<zone>/instances/<instance>', not %s", instanceURI)
+	}
+	derivedVars.project = m["project"]
+	derivedVars.zone = m["zone"]
+	derivedVars.instanceName = m["instance"]
+	return nil
+}
+
+func validateInstance(derivedVars *derivedVars, sourceOS string) error {
+	inst, err := computeClient.GetInstance(derivedVars.project, derivedVars.zone, derivedVars.instanceName)
+	if err != nil {
+		return daisy.Errf("Failed to get instance: %v", err)
+	}
+
+	if len(inst.Disks) == 0 {
+		return daisy.Errf("No disks attached to the instance.")
+	}
+	if err := validateOSDisk(inst.Disks[0], derivedVars); err != nil {
+		return err
+	}
+	if err := validateLicense(inst.Disks[0], sourceOS); err != nil {
+		return err
+	}
+
+	if inst.Metadata != nil && inst.Metadata.Items != nil {
+		for _, metadataItem := range inst.Metadata.Items {
+			if metadataItem.Key == metadataKeyWindowsStartupScriptURL {
+				derivedVars.windowsStartupScriptURLBackup = metadataItem.Value
+			} else if metadataItem.Key == metadataKeyWindowsStartupScriptURLBackup {
+				derivedVars.windowsStartupScriptURLBackupExists = true
+			}
+		}
+	}
+	// If script url backup exists, don't backup again to avoid overwriting
+	if derivedVars.windowsStartupScriptURLBackupExists {
+		derivedVars.windowsStartupScriptURLBackup = nil
```

Comment:
This feels magic -- can you think of a more explicit way to handle this control flow, other than using a mutable value?

Reply:
I explained details in validators.go line 165 in the latest commit. I  the latest commit, the logic here is not complex anymore.